### PR TITLE
stitch sources of refs that are removed too

### LINKED ISSUE #10348: Issues with updating entity relationships since upgrade to knex 1.0.2 / backstage@v0.70.0
Ever since the upgrade to knex 1.0.2, we have seen some inconsistencies with catalog data being kept up to date when the source data files are changed.

We use `@vscode/sqlite3` for local dev, and `postgres` for deployed environments. 
We are seeing the issues with both types of DB. I've even deleted the DBs (both types) and restarted everything from scratch.
I've also tested out locally with `better-sqlite3` and it has issues too.

## Expected Behavior

* When a user entities' `memberOf` groups are changed, the Group's `relations.hasMember` should reflect the changes in the DB (ie remove user from old groups and add to new groups)
* When a user entity is deleted, the Group's `relations.hasMember` should reflect the changes in the DB
* When a component's `owner` group changes, the Group's `relations.component` should reflect the changes in the DB (ie remove component from old group and add to new group)
* When a component entity is deleted, the Group's `relations.component` should reflect the changes in the DB

NOTE:
* entity deletions are done in our custom `S3Processor` via this command `this.catalogClient.removeEntityByUid(entityUid, { token: process.env.BACKEND_SECRET })`

## Actual Behavior

To confirm the issue is consistent, I've followed the following steps on both sqlite and postgres DB types and on a couple of different environments, and get the same results:
* ✅  initial load of data from entity source files into DB is accurate
* ✅  add a new user that's not in a group (user data is loaded into the DB successfully)
* ✅  add a new user that is a `memberOf` group(s) (user data is loaded and successfully added to the group's relations)
* ✅  update an existing user's metadata annotations
* 🟥  move a user's `memberOf` from one group to another (the user is added to the new group, but they're also still listed under the relations of the old group)
* 🟥  move a component's `owner` from one group to another (the component is added to the new group, but it's also still listed under the relations of the old group)
* ✅  delete a user entity that's not a member of a group (successfully deletes the user entity from the DB)
* 🟥  delete a user entity that's a member of a group (successfully deletes the user entity from the DB, but it's still listed under the `relations` of the group)
* 🟥  delete a component entity that's owned by a group (successfully deletes the component entity from the DB, but it's still listed under the `relations` of the group)

## Steps to Reproduce

As above for your favourite entity source data and Processor type

We tried the above steps with our own in house `S3Processor` and with copying the files locally and using the standard [FileReaderProcessor](https://github.com/backstage/backstage/blob/master/plugins/catalog-backend/src/modules/core/FileReaderProcessor.ts)

## Context

We noticed that several of our source data changes were not being accurately reflected in backstage after we upgraded to `knex: v1.0.2`

It looks like the old "group" relations are not kept in sync when the user `memberOf` & component `owner` fields are updated or the entity is deleted.

## Your Environment

Local dev: we use currently `"@vscode/sqlite3": "^5.0.7"`, though I've done a quick test with `better-sqlite3` and it still has issues with data updates

Deployed environments (qa/production/etc): we use `"pg": "^8.5.1",`

### PR DESCRIPTION
**Pull request by freben**
Fixes #10348

### CODE REVIEW

File: plugins/catalog-backend/src/processing/DefaultCatalogProcessingEngine.test.ts [hunk 1]
```diff
@@ -319,4 +335,115 @@ describe('DefaultCatalogProcessingEngine', () => {
 
     await engine.stop();
   });
+
+  it('should stitch both the previous and new sources when relations change', async () => {
+    const engine = new DefaultCatalogProcessingEngine(
```
Comment (freben) [inline plugins/catalog-backend/src/processing/DefaultCatalogProcessingEngine.test.ts:340]: note to self: blech, all of the tests in this file mock WAY too much and should work against a test database instead

File: .changeset/breezy-schools-compete.md [hunk 2]
```diff
@@ -0,0 +1,8 @@
+---
+'@backstage/plugin-catalog-backend': patch
+---
+
+Fixes #10348
```
Comment (Rugvip) [inline .changeset/breezy-schools-compete.md:5]: This isn't very helpful in a lot of places? 😅 Perhaps better to just describe what the issue was instead

File: plugins/catalog-backend/src/database/DefaultProcessingDatabase.ts [hunk 3]
```diff
@@ -108,9 +108,20 @@ export class DefaultProcessingDatabase implements ProcessingDatabase {
...
+    let previousRelationRows: DbRelationsRow[];
+    if (tx.client.config.client.includes('sqlite3')) {
+      previousRelationRows = await tx<DbRelationsRow>('relations')
+        .select('*')
+        .where({ originating_entity_id: id });
+      await tx<DbRelationsRow>('relations')
+        .where({ originating_entity_id: id })
+        .delete();
+    } else {
+      previousRelationRows = await tx<DbRelationsRow>('relations')
```
Comment (Rugvip) [inline plugins/catalog-backend/src/database/DefaultProcessingDatabase.ts:120]: There has to be a race here right? ^_> Thinking one entity having its relations updated, with a related one being stitched at the same time. The other one might read the relations first before deletion, but still winning the stitching afterwards.

File: .changeset/breezy-schools-compete.md (hunk 2 above)
Comment (freben) [reply .changeset/breezy-schools-compete.md:5]: Oh wut I never committed the updated text, will fix

**Review by Rugvip — APPROVED** [review]

**Pull request merged**